File: app/src/main/java/com/recursia/popularmovies/presentation/presenters/DetailScreenPresenter.java
package com.recursia.popularmovies.presentation.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.recursia.popularmovies.domain.DetailScreenInteractor;
import com.recursia.popularmovies.domain.models.Movie;
import com.recursia.popularmovies.domain.models.Trailer;
import com.recursia.popularmovies.presentation.views.DetailScreenContract;
import com.recursia.popularmovies.utils.LangUtils;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

@InjectViewState
public class DetailScreenPresenter extends MvpPresenter<DetailScreenContract> {
    private static final int MOVIE_RECOMMENDATION_PAGE = 1;
    private final DetailScreenInteractor detailScreenInteractor;
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    private final int movieId;

    public DetailScreenPresenter(DetailScreenInteractor detailScreenInteractor,
                                 int movieId) {
        this.detailScreenInteractor = detailScreenInteractor;
        this.movieId = movieId;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().hideFavoriteIcon();
        getViewState().hideMovieDetail();
        getViewState().hideRecommendationMovies();
        initData();
    }

    private void initData() {
        loadMovie();
        loadMovieRecommendations();
    }

    private void loadMovie() {
        Disposable d = detailScreenInteractor
                .getMovieById(movieId, LangUtils.getDefaultLanguage())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        this::handleMovie,
                        (t) -> getViewState().showErrorMessage(t.getLocalizedMessage())
                );
        compositeDisposable.add(d);
    }

    private void loadMovieRecommendations() {
        Disposable d = detailScreenInteractor
                .getMovieRecommendations(movieId, MOVIE_RECOMMENDATION_PAGE, LangUtils.getDefaultLanguage())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe((r) -> {
                            getViewState().setRecommendationMovies(r);
                            getViewState().showRecommendationMovies();
                        },
                        (t) -> getViewState().showErrorMessage(t.getLocalizedMessage())
                );
        compositeDisposable.add(d);
    }

    private void handleMovie(Movie movie) {
        getViewState().setMovieDetail(movie);
        getViewState().showMovieDetail();
        if (movie.isFavorite()) {
            getViewState().setFavoriteIconOn();
        } else {
            getViewState().setFavoriteIconOff();
        }
        getViewState().showFavoriteIcon();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        compositeDisposable.dispose(); //our activity is destroyed, disposing all subscriptions
    }

    public void onFavoriteIconClicked(Movie movie) {
        if (movie.isFavorite()) {
            removeMovieFavorite(movie);
        } else {
            makeMovieFavorite(movie);
        }

    }

    private void removeMovieFavorite(Movie movie) {
        Disposable d = detailScreenInteractor
                .removeFavoriteMovie(movie)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        () -> {
                            getViewState().setFavoriteIconOff();
                            getViewState().showMovieRemovedMessage();
                        }
                );
        compositeDisposable.add(d);
    }

    private void makeMovieFavorite(Movie movie) {
        Disposable d = detailScreenInteractor
                .makeFavoriteMovie(movie)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        () -> {
                            getViewState().setFavoriteIconOn();
                            getViewState().showMovieAddedMessage();
                        }
                );
        compositeDisposable.add(d);
    }

    public void onShareIconClicked(Movie movie) {
        getViewState().shareMovie(movie);
    }

    public void onMovieClicked(Movie movie) {
        getViewState().openDetailScreen(movie);
    }

    public void onBackdropImageClicked(Movie movie) {
        getViewState().openPhotoDetail(movie.getBackdropPath());
    }

    public void onTrailerPlayButtonClicked(Trailer trailer) {
        getViewState().openTrailerUrl(trailer);
    }

}

File: app/src/main/java/com/recursia/popularmovies/presentation/presenters/DetailScreenPresenter.kt
package com.recursia.popularmovies.presentation.presenters

import com.recursia.popularmovies.Screens
import com.recursia.popularmovies.domain.DetailScreenInteractor
import com.recursia.popularmovies.domain.models.Genre
import com.recursia.popularmovies.domain.models.Movie
import com.recursia.popularmovies.domain.models.enums.MovieStatus
import com.recursia.popularmovies.presentation.views.contracts.DetailScreenContract
import com.recursia.popularmovies.utils.LangUtils
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import moxy.InjectViewState
import moxy.MvpPresenter
import ru.terrakok.cicerone.Router

/**
 * Copyright Alexander Silinsky 2020
 * Date 10.04.2020
 * Detail screen presenter
 */
@InjectViewState
class DetailScreenPresenter(
        private val detailScreenInteractor: DetailScreenInteractor, //detail screen interactor
        private val router: Router, // router for navigating
        private val movieId: Int // movie id
) : MvpPresenter<DetailScreenContract>() {
    private val compositeDisposable = CompositeDisposable()

    /**
     * Calls on first view attach
     */
    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        initData()
    }

    /**
     * Function to init data
     */
    private fun initData() {
        val d = detailScreenInteractor
                .getMovieById(movieId, LangUtils.defaultLanguage)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { viewState.setMovieDetail(it) },
                        { viewState.showErrorMessage(it.localizedMessage) }
                )
        compositeDisposable.add(d)
    }

    /**
     * Calls on destroy
     */
    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }

    /**
     * Calls on back pressed callback
     */
    fun onBackPressed() {
        router.exit()
    }

    /**
     * On share icon clicked callback
     * @param movie movie to share
     */
    fun onShareIconClicked(movie: Movie?) {
        movie?.let {
            viewState.shareMovie(movie)
        }
    }

    /**
     * On movie status clicked callback
     * save movie to user collection
     * @param movie movie
     * @param status status
     */
    fun onMovieStatusClicked(movie: Movie?, status: MovieStatus) {
        movie?.let {
            val d = detailScreenInteractor
                    .setMovieStatus(movie)
                    .observeOn(AndroidSchedulers.mainThread())
                    .doOnSubscribe { movie.status = status }
                    .subscribe(
                            { viewState.setMovieStatus(status) },
                            { viewState.showErrorMessage(it.localizedMessage) }
                    )
            compositeDisposable.add(d)
        }
    }

    /**
     * On backdrop clicked callback
     * @param movie movie
     */
    fun onBackdropClicked(movie: Movie?) {
        movie?.let {
            router.navigateTo(Screens.PhotoScreen(movie.posterPath!!))
        }
    }

    /**
     * On genre clicked callback
     * @param genre genre
     */
    fun onGenreClicked(genre: Genre) {
        viewState.showBottomSheetGenreMovies(genre)
    }
}
